In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [2]:
dt = pd.read_csv("C:/Users/madha/Downloads/Customer Churn assessment/customer_churn_large_dataset_csv.csv")
dt

CustomerID             Name  Age  Gender     Location  \
0               1       Customer_1   63    Male  Los Angeles   
1               2       Customer_2   62  Female     New York   
2               3       Customer_3   24  Female  Los Angeles   
3               4       Customer_4   36  Female        Miami   
4               5       Customer_5   46  Female        Miami   
...           ...              ...  ...     ...          ...   
99995       99996   Customer_99996   33    Male      Houston   
99996       99997   Customer_99997   62  Female     New York   
99997       99998   Customer_99998   64    Male      Chicago   
99998       99999   Customer_99999   51  Female     New York   
99999      100000  Customer_100000   27  Female  Los Angeles   

       Subscription_Length_Months  Monthly_Bill  Total_Usage_GB  Churn  
0                              17         73.36             236      0  
1                               1         48.76             172      0  
2                               5         85.47             460      0  
3                               3         97.94             297      1  
4                              19         58.14             266      0  
...                           ...           ...             ...    ...  
99995                          23         55.13             226      1  
99996                          19         61.65             351      0  
99997                          17         96.11             251      1  
99998                          20         49.25             434      1  
99999                          19         76.57             173      1  

[100000 rows x 9 columns]

In [3]:
dt.columns

Index(['CustomerID', 'Name', 'Age', 'Gender', 'Location',
       'Subscription_Length_Months', 'Monthly_Bill', 'Total_Usage_GB',
       'Churn'],
      dtype='object')

In [4]:
dt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 9 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   CustomerID                  100000 non-null  int64  
 1   Name                        100000 non-null  object 
 2   Age                         100000 non-null  int64  
 3   Gender                      100000 non-null  object 
 4   Location                    100000 non-null  object 
 5   Subscription_Length_Months  100000 non-null  int64  
 6   Monthly_Bill                100000 non-null  float64
 7   Total_Usage_GB              100000 non-null  int64  
 8   Churn                       100000 non-null  int64  
dtypes: float64(1), int64(5), object(3)
memory usage: 6.9+ MB


In [5]:
dt.isnull()

CustomerID   Name    Age  Gender  Location  Subscription_Length_Months  \
0           False  False  False   False     False                       False   
1           False  False  False   False     False                       False   
2           False  False  False   False     False                       False   
3           False  False  False   False     False                       False   
4           False  False  False   False     False                       False   
...           ...    ...    ...     ...       ...                         ...   
99995       False  False  False   False     False                       False   
99996       False  False  False   False     False                       False   
99997       False  False  False   False     False                       False   
99998       False  False  False   False     False                       False   
99999       False  False  False   False     False                       False   

       Monthly_Bill  Total_Usage_GB  Churn  
0             False           False  False  
1             False           False  False  
2             False           False  False  
3             False           False  False  
4             False           False  False  
...             ...             ...    ...  
99995         False           False  False  
99996         False           False  False  
99997         False           False  False  
99998         False           False  False  
99999         False           False  False  

[100000 rows x 9 columns]

In [6]:
dt.isnull().sum()

CustomerID                    0
Name                          0
Age                           0
Gender                        0
Location                      0
Subscription_Length_Months    0
Monthly_Bill                  0
Total_Usage_GB                0
Churn                         0
dtype: int64

In [7]:
dt.shape

(100000, 9)

In [8]:
dt["Churn"].unique()

array([0, 1], dtype=int64)

In [9]:
dt["Churn"].value_counts()

0    50221
1    49779
Name: Churn, dtype: int64

-> There is no irregularity in this dataset, so we don't have to perform the data handling process.

In [10]:
dt[dt.isnull()]

CustomerID Name  Age Gender Location  Subscription_Length_Months  \
0             NaN  NaN  NaN    NaN      NaN                         NaN   
1             NaN  NaN  NaN    NaN      NaN                         NaN   
2             NaN  NaN  NaN    NaN      NaN                         NaN   
3             NaN  NaN  NaN    NaN      NaN                         NaN   
4             NaN  NaN  NaN    NaN      NaN                         NaN   
...           ...  ...  ...    ...      ...                         ...   
99995         NaN  NaN  NaN    NaN      NaN                         NaN   
99996         NaN  NaN  NaN    NaN      NaN                         NaN   
99997         NaN  NaN  NaN    NaN      NaN                         NaN   
99998         NaN  NaN  NaN    NaN      NaN                         NaN   
99999         NaN  NaN  NaN    NaN      NaN                         NaN   

       Monthly_Bill  Total_Usage_GB  Churn  
0               NaN             NaN    NaN  
1               NaN             NaN    NaN  
2               NaN             NaN    NaN  
3               NaN             NaN    NaN  
4               NaN             NaN    NaN  
...             ...             ...    ...  
99995           NaN             NaN    NaN  
99996           NaN             NaN    NaN  
99997           NaN             NaN    NaN  
99998           NaN             NaN    NaN  
99999           NaN             NaN    NaN  

[100000 rows x 9 columns]

In [11]:
rows_with_null = dt.isnull().any(axis=1)
count_null_rows = rows_with_null.sum()

In [12]:
dt[rows_with_null]

Empty DataFrame
Columns: [CustomerID, Name, Age, Gender, Location, Subscription_Length_Months, Monthly_Bill, Total_Usage_GB, Churn]
Index: []

In [13]:
count_null_rows

0

**Feature Engineering**
* removing "CustomerID" & "Name" column because they are not necessary to performe required taks
* changing datatype of Gender column by  replacing Male by 1 and Female by 0

In [14]:
dt1=dt.drop(columns=["CustomerID", "Name"])

In [15]:
dt1['Gender'].replace({'Male':1,'Female':0},inplace=True)
dt1

Age  Gender     Location  Subscription_Length_Months  Monthly_Bill  \
0       63       1  Los Angeles                          17         73.36   
1       62       0     New York                           1         48.76   
2       24       0  Los Angeles                           5         85.47   
3       36       0        Miami                           3         97.94   
4       46       0        Miami                          19         58.14   
...    ...     ...          ...                         ...           ...   
99995   33       1      Houston                          23         55.13   
99996   62       0     New York                          19         61.65   
99997   64       1      Chicago                          17         96.11   
99998   51       0     New York                          20         49.25   
99999   27       0  Los Angeles                          19         76.57   

       Total_Usage_GB  Churn  
0                 236      0  
1                 172      0  
2                 460      0  
3                 297      1  
4                 266      0  
...               ...    ...  
99995             226      1  
99996             351      0  
99997             251      1  
99998             434      1  
99999             173      1  

[100000 rows x 7 columns]

In [16]:
dt2 = pd.get_dummies(data=dt1, columns=['Location'])
dt2.columns

Index(['Age', 'Gender', 'Subscription_Length_Months', 'Monthly_Bill',
       'Total_Usage_GB', 'Churn', 'Location_Chicago', 'Location_Houston',
       'Location_Los Angeles', 'Location_Miami', 'Location_New York'],
      dtype='object')

In [17]:
dt2.dtypes

Age                             int64
Gender                          int64
Subscription_Length_Months      int64
Monthly_Bill                  float64
Total_Usage_GB                  int64
Churn                           int64
Location_Chicago                uint8
Location_Houston                uint8
Location_Los Angeles            uint8
Location_Miami                  uint8
Location_New York               uint8
dtype: object

**Model Building**
* In this phase we are traning and testing our dataset.
* After that we are going to apply our machine learning algorithm

In [18]:
X = dt2.drop('Churn', axis=1)
Y = dt2['Churn']
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [19]:
X.shape

(100000, 10)

In [20]:
Y.shape

(100000,)

In [21]:
X_train.shape

(80000, 10)

In [74]:
X_test.shape

(20000, 10)

In [75]:
X_train

Age  Gender  Subscription_Length_Months  Monthly_Bill  Total_Usage_GB  \
75220   54       0                           5         84.50             205   
48955   28       1                          24         82.06             239   
44966   57       1                          12         52.29              62   
13568   19       1                          19         32.57             173   
92727   56       0                           8         33.52             314   
...    ...     ...                         ...           ...             ...   
6265    35       1                          21         67.33             235   
54886   56       1                          13         85.40             347   
76820   69       1                           2         76.24             321   
860     55       1                          12         89.19             315   
15795   26       0                          17         70.41             335   

       Location_Chicago  Location_Houston  Location_Los Angeles  \
75220                 0                 0                     0   
48955                 0                 0                     0   
44966                 1                 0                     0   
13568                 0                 1                     0   
92727                 0                 0                     0   
...                 ...               ...                   ...   
6265                  0                 0                     0   
54886                 1                 0                     0   
76820                 0                 1                     0   
860                   1                 0                     0   
15795                 0                 0                     1   

       Location_Miami  Location_New York  
75220               0                  1  
48955               0                  1  
44966               0                  0  
13568               0                  0  
92727               1                  0  
...               ...                ...  
6265                1                  0  
54886               0                  0  
76820               0                  0  
860                 0                  0  
15795               0                  0  

[80000 rows x 10 columns]

In [78]:
Y_test

75721    0
80184    0
19864    0
76699    1
92991    0
        ..
32595    0
29313    1
37862    1
53421    0
42410    1
Name: Churn, Length: 20000, dtype: int64

* Now we apply logistic regression on our trained dataset

In [55]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(X_train, Y_train)

LogisticRegression()

In [56]:
Y_pred = model.predict(X_test)
Y_pred

array([0, 0, 0, ..., 1, 0, 1], dtype=int64)

In [57]:
model.score(X,Y)

0.50203

In [58]:

from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
accuracy = accuracy_score(Y_test, Y_pred)
precision = precision_score(Y_test, Y_pred)
recall = recall_score(Y_test, Y_pred)
f1 = f1_score(Y_test, Y_pred)
accuracy

0.49995

In [59]:
precision

0.49323638823131555

In [60]:
recall

0.2940227799616974

In [61]:
f1

0.3684243763814335

**Model Optimization**
* After logistic regression model result we are going to check your data on new model.
* The new model we are using in this stage is called Decision Tree Classifier

In [69]:
from sklearn.tree import DecisionTreeClassifier

In [70]:
model2 = DecisionTreeClassifier()

In [71]:
model2.fit(X,Y)

DecisionTreeClassifier()

In [72]:
model2.score(X,Y)

1.0

In [73]:
Y_predict=model2.predict(X_test)
Y_predict

array([0, 0, 0, ..., 1, 0, 1], dtype=int64)

**Result**
* So after applying both model on our dataset our results are:
Model score (Logistic regression) = 0.50203 /
Model score (DecisionTreeClassifier) = 1.0
* Here we can see that DecisionTreeClassifier has better score than logistic regression

**Model Deployment**

In [68]:
#user_age = input("Please enter age: ")
#user_gender = input("Please enter gender: ")
#user_location = input("Please enter location: ")
#user_subscription = input("Please enter subscription: ")
#user_Mbill = input("Please enter Mbill: ")
#user_TuGB = input("Please enter TuGB: ")
user_input = []

feature_names = ["Age", "Gender", "Subscription_Length_Months", "Monthly_Bill", "Total_Usage_GB", "Location_Chicago", "Location_Houston", "Location_Los Angeles", "Location_Miami", "Location_New York"]

for feature_name in feature_names:
    value = float(input(f"Enter {feature_name}: "))
    user_input.append(value)

# Step 2: Preprocess user input (if needed)
# You might need to scale, normalize, or preprocess the user input based on your model's requirements.

# Step 3: Make predictions
user_input = np.array(user_input).reshape(1, -1)  # Reshape input for the model
prediction = model.predict(user_input)

Enter Age: 68
Enter Gender: 1
Enter Subscription_Length_Months: 11
Enter Monthly_Bill: 30.91
Enter Total_Usage_GB: 348
Enter Location_Chicago: 1
Enter Location_Houston: 0
Enter Location_Los Angeles: 0
Enter Location_Miami: 0
Enter Location_New York: 0


C:\Users\madha\anaconda3\lib\site-packages\sklearn\base.py:420: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
